File: ios/WidgetViewController.m
//
//  WidgetViewController.m
//  TodayWidgetDemo
//
//  Created by Telldus Technologies AB on 28/05/18.
//  Copyright © 2018 Telldus Technologies AB. All rights reserved.
//

#import "WidgetViewController.h"
#import "Widget.h"

@interface WidgetViewController ()
@property NSMutableArray<Widget*> *devices;
//@property NSMutableData *responseData;
@property Widget *selectedDevice;
@property UIPickerView *picker;
@property NSString* accessToken;
@end

@implementation WidgetViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  
  _accessToken = [[[NSUserDefaults alloc] initWithSuiteName:@"group.com.telldus.TokenUserDefaults"] stringForKey:@"accessToken"];
  
  _devices = [[NSMutableArray alloc] init];
  
  UIToolbar *toolBar = [[UIToolbar alloc] init];
  toolBar.barStyle = UIBarStyleDefault;
  [toolBar setTranslucent:YES];
  toolBar.tintColor = [UIColor colorWithRed:76/255 green:217/255 blue:100/255 alpha:1];
  [toolBar sizeToFit];
  
  UIBarButtonItem *doneBtn = [[UIBarButtonItem alloc] initWithTitle:@"Done" style:UIBarButtonItemStylePlain target:self action:@selector(donePicker)];
  UIBarButtonItem *flexibleItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace target:self action:nil];
  UIBarButtonItem *cancelBtn = [[UIBarButtonItem alloc] initWithTitle:@"Cancel" style:UIBarButtonItemStylePlain target:self action:@selector(cancelPicker)];
  NSArray *array = @[cancelBtn, flexibleItem, doneBtn];
  [toolBar setItems: array];
  [toolBar setUserInteractionEnabled:YES];
  
  _picker = [UIPickerView new];
  _picker.delegate = self;
  _picker.dataSource = self;
  _picker.showsSelectionIndicator = YES;
  _selectDevices.inputView = _picker;
  _selectDevices.inputAccessoryView = toolBar;
  
  [self getAllDevices];  
  _devicefontimage.text = @"device-alt";
  _devicefontimage.textColor = UIColor.whiteColor;
 
}

- (void)viewWillAppear:(BOOL)animated {
  [super viewWillAppear:animated];
 [_navBar setBarTintColor:[UIColor colorWithRed:0 green:0 blue:0 alpha:0]];
  
  UIImageView *imageView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, 40, 40)];
  [imageView setContentMode:UIViewContentModeScaleAspectFit];
  imageView.image=[UIImage imageNamed:@"telldus_1.png"];
  [self.navigationItem.titleView addSubview:imageView];
}

-(void)donePicker {
  [self.view endEditing:YES];
  _selectDevices.text = _selectedDevice.name;
  WidgetDA *widgetDa = [[WidgetDA alloc] init];
  
//  NSLog()
  
  if ((_selectedDevice.state == 1) || (_selectedDevice.state == 2)) {
    _selectedDevice.devType = @"DEVICE";
  } else if (_selectedDevice.state == 4) {
    _selectedDevice.devType = @"BELL";
  } else if (_selectedDevice.state == 16) {
    _selectedDevice.devType = @"DIM";
  } else if (_selectedDevice.state == 32) {
    _selectedDevice.devType = @"LEARN";
  } else if (_selectedDevice.state == 128 || _selectedDevice.state == 256 || _selectedDevice.state == 512) {
    _selectedDevice.devType = @"CONTROLDEV";
  }
  [widgetDa addDevice:_selectedDevice];
}

-(void)cancelPicker {
  [self.view endEditing:YES];
}

- (void)getAllDevices {
  
  NSLog(@"--->>%@", _accessToken);
  NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString: @"https://api3.telldus.com/oauth2/devices/list?supportedMethods=951&includeIgnored=1"] cachePolicy:NSURLRequestUseProtocolCachePolicy timeoutInterval:60.0];
  [request setHTTPMethod:@"GET"];  
  [request setValue:_accessToken forHTTPHeaderField:@"Authorization"];
  [request setValue:@"application/json" forHTTPHeaderField:@"Accept"];
  [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
  
  NSURLResponse * response = nil;
  NSError * error = nil;
  NSData * data = [NSURLConnection sendSynchronousRequest:request
                                        returningResponse:&response
                                                    error:&error];
  if (error == nil) {
    NSError *error1;
    NSMutableDictionary * innerJson = [NSJSONSerialization
                                       JSONObjectWithData:data options:kNilOptions error:&error1
                                       ];
    NSLog(@"response json: %@", innerJson);
    NSMutableArray *array = innerJson[@"device"];
    for (NSDictionary *dict in array) {
      Widget *widget = [[Widget alloc] init];
      widget.name = dict[@"name"];
      widget.clientDeviceId = dict[@"clientDeviceId"];
      widget.type = dict[@"type"];
      widget.client = dict[@"client"];
      widget.identifier = dict[@"id"];
      widget.clientName = dict[@"clientName"];
      widget.metadataHash = dict[@"metadataHash"];
      widget.statevalue = (dict[@"statevalue"] == nil || dict[@"statevalue"] == [NSNull null]) ? @"" : dict[@"statevalue"];
      widget.state = [dict[@"state"] intValue];
      widget.ignored = [dict[@"ignored"] intValue];
      widget.methods = [dict[@"methods"] intValue];
      widget.editable = [dict[@"editable"] intValue];
      widget.online = [dict[@"online"] intValue];
      [self.devices addObject:widget];
    }
    
    if ([_devices count] > 0) {
      _selectedDevice = [self.devices objectAtIndex:0];
    }
  } else {
    NSLog(@"Request failed %@", error.localizedDescription);
  }
}

- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
  return 1;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
  return _devices.count;
}

- (NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component {
  return [self.devices objectAtIndex:row].name;
}

- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component {
  if(_devices.count > 0) {
    _selectedDevice = [self.devices objectAtIndex:row];
  }
}
@end

File: ios/TelldusDeviceWidget/WidgetViewController.swift
//
//  WidgetViewController.swift
//  Devices
//

import UIKit
import Alamofire
import RealmSwift
import SwiftyJSON

class WidgetViewController: UIViewController, UIPickerViewDelegate, UIPickerViewDataSource {
    
    var realm = try? Realm()
    
    let widgetPicker = UIPickerView()
    var devices: [JSON] = []
    
    
    
    
    @IBOutlet weak var addDevLblView: UIView!
    @IBOutlet weak var selectDevices: UITextField!
    @IBOutlet weak var cnfrmBtmLbl: UIButton!
    @IBOutlet weak var cnclBtnLbl: UIButton!
    @IBOutlet weak var navBar: UINavigationBar!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        print("---loading")
        
        addDevLblView.backgroundColor = UIColor(patternImage: #imageLiteral(resourceName: "telldus_geometric_header_bg"))
        
        cnfrmBtmLbl.layer.cornerRadius = 20
        cnclBtnLbl.layer.cornerRadius = 20
        
        widgetPicker.showsSelectionIndicator = true
        widgetPicker.delegate = self
        widgetPicker.dataSource = self
        
        let toolBar = UIToolbar()
        toolBar.barStyle = UIBarStyle.default
        toolBar.isTranslucent = true
        toolBar.tintColor = UIColor(red: 76/255, green: 217/255, blue: 100/255, alpha: 1)
        toolBar.sizeToFit()
        
        let doneButton = UIBarButtonItem(title: "Done", style: UIBarButtonItemStyle.plain, target: self, action: #selector(WidgetViewController.saveAction))
        let spaceButton = UIBarButtonItem(barButtonSystemItem: UIBarButtonSystemItem.flexibleSpace, target: nil, action: nil)
        let cancelButton = UIBarButtonItem(title: "Cancel", style: UIBarButtonItemStyle.plain, target: self, action: #selector(WidgetViewController.cancelAction))
        
        toolBar.setItems([cancelButton, spaceButton, doneButton], animated: false)
        toolBar.isUserInteractionEnabled = true
        
        selectDevices.inputView = widgetPicker
        selectDevices.inputAccessoryView = toolBar
        
        let directory: URL = FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: K_GROUP_ID)!
        
        let fileURL = directory.appendingPathComponent(K_DB_NAME)
        realm = try! Realm(fileURL: fileURL)
        
        getAllDevices()
    }
    
    override func viewDidAppear(_ animated: Bool) {
        self.navBar.barTintColor = UIColor(red: 0, green: 0, blue: 128, alpha: 0)
        
        let imageView = UIImageView(frame: CGRect(x: 0, y: 0, width: 40, height: 40))
        imageView.contentMode = .scaleAspectFit
        
        let image = UIImage(named: "telldus_1.png")
        imageView.image = image
        
        navigationItem.titleView = imageView
    }
    
    func getAllDevices() {
        
        let headers: HTTPHeaders = [
            "Authorization": "Bearer 7ab5de2d912cc354d0888b17bd79a78ef4c784a0",
            "Accept": "application/json",
            "Content-Type": "application/json"
        ]
        
        Alamofire.request("https://api3.telldus.com/oauth2/devices/list?supportedMethods=951&includeIgnored=1", method: .get, headers: headers).responseJSON { response in
            switch response.result {
            case .success:
                if response.data != nil {
                    let json =  try? JSON(data: response.data!)
                    
                    if let allDevices = json!["device"].array {
                        print (allDevices)
                        print (allDevices.count)
                        for device in allDevices {
                            self.devices.append(device)
                        }
                    }
                }
                
            case .failure(let error):
                print(error)
            }
        }
    }
    
    @objc func saveAction() {
        
        let widget = Widget()
        
        for device in devices {
            
            
            if let title = device["name"].string {
                if(title == selectDevices.text) {
                    if let selDev = device.dictionary {
                        widget.state = selDev["state"]!.int!
                        widget.name = title
                        widget.ignored = selDev["ignored"]!.int!
                        widget.clientDeviceId = selDev["clientDeviceId"]!.string!
                        widget.type = selDev["type"]!.string!
                        widget.client = selDev["client"]!.string!
                        widget.methods = selDev["methods"]!.int!
                        widget.id = selDev["id"]!.string!
                        widget.editable = selDev["editable"]!.int!
                        widget.clientName = selDev["clientName"]!.string ?? ""
                        widget.metadataHash = selDev["metadataHash"]!.string ?? ""
                        widget.statevalue = selDev["statevalue"]!.string ?? ""
                        if let online = device["online"].string {
                            widget.online = Int(online)!
                        } else {
                            widget.online = selDev["online"]!.int!
                        }
                        if (selDev["state"]!.int! == 1 || selDev["state"]!.int! == 2) {
                            widget.devType = "DEVICE"
                        } else if (selDev["state"]!.int! == 4) {
                            widget.devType = "BELL"
                        } else if (selDev["state"]!.int! == 16) {
                            widget.devType = "DIM"
                        } else if (selDev["state"]!.int! == 32) {
                            widget.devType = "LEARN"
                        } else if (selDev["state"]!.int! == 128 || selDev["state"]!.int! == 256 || selDev["state"]!.int! == 512) {
                            widget.devType = "CONTROLDEV"
                        }
                    }
                }
            }
        }
        
        try! realm?.write {
            realm?.add(widget)
        }
        
        self.view.endEditing(true)
    }
    
    @objc func cancelAction() {
        self.view.endEditing(true)
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return 1
    }
    
    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        return devices.count
    }
    
    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        
        if let title = devices[row]["name"].string {
            return title
        } else {
            return ""
        }
        
    }
    
    func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
        if(devices.count > 0) {
            if let title = devices[row]["name"].string {
                selectDevices.text = title
            } else {
                selectDevices.text = ""
            }
        }
    }
}

extension String {
    func toJSON() -> Any? {
        guard let data = self.data(using: .utf8, allowLossyConversion: false) else { return nil }
        return try? JSONSerialization.jsonObject(with: data, options: .mutableContainers)
    }
}
